selfservice/hook/web_hook.go

```diff
@@ -0,0 +1,250 @@
+	"":           newNoopAuthStrategy,
+	"api-key":    newApiKeyStrategy,
+	"basic-auth": newBasicAuthStrategy,
+}
+
+func newAuthStrategy(name string, c json.RawMessage) (as AuthStrategy, err error) {
+	if f, ok := strategyFactories[name]; ok {
+		as, err = f(c)
+	} else {
+		err = fmt.Errorf("unsupported auth type: %s", name)
+	}
+	return
+}
+
+func newNoopAuthStrategy(_ json.RawMessage) (AuthStrategy, error) {
+	return &noopAuthStrategy{}, nil
+}
+
+func (c *noopAuthStrategy) apply(_ *http.Request) {}
+
+func newBasicAuthStrategy(raw json.RawMessage) (AuthStrategy, error) {
+	type config struct {
+		User     string
+		Password string
+	}
+
+	var c config
+	if err := json.Unmarshal(raw, &c); err != nil {
+		return nil, err
+	}
+
+	return &basicAuthStrategy{
+		user:     c.User,
+		password: c.Password,
+	}, nil
+}
+
+func (c *basicAuthStrategy) apply(req *http.Request) {
+	req.SetBasicAuth(c.user, c.password)
+}
+
+func newApiKeyStrategy(raw json.RawMessage) (AuthStrategy, error) {
+	type config struct {
+		In    string
+		Name  string
+		Value string
+	}
+
+	var c config
+	if err := json.Unmarshal(raw, &c); err != nil {
+		return nil, err
+	}
+
+	return &apiKeyStrategy{
+		in:    c.In,
+		name:  c.Name,
+		value: c.Value,
+	}, nil
+}
+
+func (c *apiKeyStrategy) apply(req *http.Request) {
+	switch c.in {
+	case "cookie":
+		req.AddCookie(&http.Cookie{Name: c.name, Value: c.value})
+	default:
+		req.Header.Set(c.name, c.value)
+	}
+}
+
+func newWebHookConfig(r json.RawMessage) (*webHookConfig, error) {
+	type rawWebHookConfig struct {
+		Method string
+		Url    string
+		Body   string
+		Auth   struct {
+			Type   string
+			Config json.RawMessage
+		}
+	}
+
+	var rc rawWebHookConfig
+	err := json.Unmarshal(r, &rc)
+	if err != nil {
+		return nil, err
+	}
+
+	as, err := newAuthStrategy(rc.Auth.Type, rc.Auth.Config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create web hook auth strategy: %w", err)
+	}
+
+	return &webHookConfig{
+		method:       rc.Method,
+		url:          rc.Url,
+		templatePath: rc.Body,
+		auth:         as,
+	}, nil
+}
+
+func NewWebHook(r webHookDependencies, c json.RawMessage) *WebHook {
+	return &WebHook{r: r, c: c}
+}
+
+func (e *WebHook) ExecutePostVerificationHook(_ http.ResponseWriter, req *http.Request, flow *verification.Flow, session *session.Session) error {
+	return e.execute(&templateContext{
+		Flow:           flow,
+		RequestHeaders: req.Header,
+		RequestMethod:  req.Method,
+		RequestUrl:     req.URL.String(),
+		Session:        session,
+	})
+}
+
+func (e *WebHook) ExecutePostRegistrationPostPersistHook(_ http.ResponseWriter, req *http.Request, flow *registration.Flow, session *session.Session) error {
+	return e.execute(&templateContext{
+		Flow:           flow,
+		RequestHeaders: req.Header,
+		RequestMethod:  req.Method,
+		RequestUrl:     req.URL.String(),
+		Session:        session,
+	})
+}
+
+func (e *WebHook) execute(data *templateContext) error {
+	conf, err := newWebHookConfig(e.c)
```

Comment:
actually yes and I would love to change this, but how should an error be properly logged and propagated during a hook creation?

I mean this is the code responsible for hook creation:

```.go
func (m *RegistryDefault) getHooks(credentialsType string, configs []config.SelfServiceHook) (i []interface{}) {
	for _, h := range configs {
		switch h.Name {
		case hook.KeySessionIssuer:
			i = append(i, m.HookSessionIssuer())
		case hook.KeySessionDestroyer:
			i = append(i, m.HookSessionDestroyer())
		case hook.KeyWebHook:
			i = append(i, hook.NewWebHook(m, h.Config))
		default:
			var found bool
			for name, m := range m.injectedSelfserviceHooks {
				if name == h.Name {
					i = append(i, m(h))
					found = true
					break
				}
			}
			if found {
				continue
			}
			m.l.
				WithField("for", credentialsType).
				WithField("hook", h.Name).
				Errorf("A unknown hook was requested and can therefore not be used")
		}
	}

	return i
}
```

I could have made `hook.NewWebHook` return an error as well, but it just didn't feel right for me just to log it here and continue.

Reply:
Got it! Makes sense.

Reply:
@harnash: After thinking a bit about your question: 
Kratos supports config reload on the fly. That way the configuration should always be parsed on every hook execution and not be cached. So whether the config is deserialized during the hook creation or execution, doesn't matter at all as the execution happens directly after the creation.

Reply:
@abador: Maybe we should take it as a TODO for the future as an enhancement of the implementation

Reply:
Sounds good!

Reply:
Yeah this could definitely be improved here. I think for now this is fine though as these hooks only run when people sign up, sign in (both of which use BCrypt so they're slow anyways) or change their settings. I think the performance impact is negligible until you hit 50% Google scale